node/header_exchange.go

```diff
@@ -0,0 +1,241 @@
+package node
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+
+	"github.com/celestiaorg/go-header"
+	goheaderp2p "github.com/celestiaorg/go-header/p2p"
+	goheaderstore "github.com/celestiaorg/go-header/store"
+	"github.com/celestiaorg/go-header/sync"
+	goheadersync "github.com/celestiaorg/go-header/sync"
+	ds "github.com/ipfs/go-datastore"
+	badger3 "github.com/ipfs/go-ds-badger3"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/p2p/net/conngater"
+	"github.com/tendermint/tendermint/libs/log"
+	tmtypes "github.com/tendermint/tendermint/types"
+	"go.uber.org/multierr"
+
+	"github.com/rollkit/rollkit/config"
+	"github.com/rollkit/rollkit/p2p"
+	"github.com/rollkit/rollkit/types"
+)
+
+type HeaderExchangeService struct {
+	conf            config.NodeConfig
+	genesis         *tmtypes.GenesisDoc
+	p2p             *p2p.Client
+	ex              *goheaderp2p.Exchange[*types.Header]
+	syncer          *sync.Syncer[*types.Header]
+	sub             *goheaderp2p.Subscriber[*types.Header]
+	p2pServer       *goheaderp2p.ExchangeServer[*types.Header]
+	headerStore     *goheaderstore.Store[*types.Header]
+	syncerStarted   bool
+	syncedHeadersCh chan *types.SignedHeader
+
+	logger log.Logger
+	ctx    context.Context
+}
+
+func NewHeaderExchangeService(ctx context.Context, ds ds.TxnDatastore, conf config.NodeConfig, genesis *tmtypes.GenesisDoc, p2p *p2p.Client, syncedHeadersCh chan *types.SignedHeader, logger log.Logger) (*HeaderExchangeService, error) {
+	// mainKV is TxnDatastore, but we require Batching, hence the type conversion
+	// note, badger datastore implements both
+	mainKVBatch, ok := ds.(*badger3.Datastore)
```

Comment:
I think we should type-assert `ds.Batching` here, to be a bit more generic, require minimal interface, and be future-proof.